File: c/pe005.c
#include <stdio.h>

#define MAX 20

unsigned int isprime(unsigned int);

int main()
{
  unsigned int i = 2;
  unsigned int try_product = 1;
  unsigned int total = 1;
  while (i <= MAX) {
    if (isprime(i)) {
      if (try_product) {
        if ((i * i) > MAX) {
          try_product = 0;
        } else {
          unsigned int tmp = i;
          unsigned int last = tmp;
          while (tmp <= MAX) {
            last = tmp;
            tmp *= i;
          }
          total *= last;
        }
      }
      if (!(try_product))
        total *= i;
    }
    i++;
  }
  printf("%d\n",total);
  return 0;
}

unsigned int isprime(unsigned int num) {
  if (num % 2) {
    if (num < 8) {
      return !(num == 1);
    } else {
      unsigned int divisor = 3;
      while ((divisor * divisor) <= num) {
        if (!(num % divisor))
          return 0;
        divisor += 2;
      }
      return 1;
    }
  } else {
    return num == 2;
  }
}

File: go/pe005.go
package main

import (
	"fmt"
)

func is_prime(num int) bool {
	if num%2 == 1 {
		if num < 8 {
			return num != 1
		} else {
			var divisor int
			divisor = 3
			for divisor*divisor <= num {
				if num%divisor == 0 {
					return false
				}
				divisor += 2
			}
			return true
		}
	} else {
		return num == 2
	}
}

func main() {
	const MAX = 20
	var i, total int
	var try_product bool

	i = 2
	total = 1
	try_product = true
	for i < MAX {
		if is_prime(i) {
			if try_product {
				if i*i > MAX {
					try_product = false
				} else {
					var tmp, last int
					tmp = i
					last = tmp
					for tmp < MAX {
						last = tmp
						tmp *= i
					}
					total *= last
				}
			}
			if !try_product {
				total *= i
			}
		}
		i += 1
	}
	fmt.Printf("%d\n", total)
}
